rfcs/20240312-standardize-tan-customcall-tuplecollective-ops.md

```diff
@@ -0,0 +1,359 @@
+defined as follows:
+
+* `cross_replica(replica_groups)`
+  if `channel_id <= 0 and use_global_device_ids = false`.
+* `cross_replica_and_partition(replica_groups)`
+  if `channel_id > 0 and use_global_device_ids = false`.
+* `flattened_ids(replica_groups)`
+  if `channel_id > 0 and use_global_device_ids = true`.
+
+Afterwards, within each process_group and for each i in [0, size(operands)]:
+
+* `operands[i]@receiver = [operand[i]@sender for sender in process_group]` for all
+  `receiver` in `process_group`.
+* `results[i]@process = concatenate(operands[i]@process, all_gather_dim)` for all
+  `process` in `process_group`.
+
+#### Inputs
+
+| Label | Name                    | Type                                         | Constraints |
+|-------|-------------------------|----------------------------------------------|-------------|
+| (I1)  | `operands`               | variadic tensors or per-tensor quantized tensors        | (C1), (C6)  |
+| (I2)  | `all_gather_dim`        | constant of type `si64`                      | (C1), (C6)  |
+| (I3)  | `replica_groups`        | 2-dimensional tensor constant of type `si64` | (C2-C4)     |
+| (I4)  | `channel_id`            | constant of type `si64`                      | (C5)        |
+| (I5)  | `use_global_device_ids` | constant of type `i1`                        | (C5)        |
+
+#### Outputs
+
+| Name     | Type                                  | Constraints |
+|----------|---------------------------------------|-------------|
+| `results` | variadic tensors or per-tensor quantized tensors | (C6)        |
+
+#### Constraints
+
+* (C1) `0 <= all_gather_dim < rank(operands...)`.
+* (C2) `is_unique(replica_groups)`.
+* (C3) `size(replica_groups)` is defined as:
+  * `num_replicas` if `cross_replica` is used.
+  * `num_replicas` if `cross_replica_and_partition` is used.
+  * `num_processes` if `flattened_ids` is used.
+* (C4) `0 <= replica_groups < size(replica_groups)`.
+* (C5) If `use_global_device_ids = true`, then `channel_id > 0`.
+* (C6) `type(results...) = type(operands...)` except:
+  * `dim(results..., all_gather_dim) =
+    dim(operands..., all_gather_dim) * dim(process_groups, 1)`.
+
+#### Examples
+
+```mlir
+// num_replicas: 2
+// num_partitions: 1
+// %operand@(0, 0): [[1, 2], [3, 4]]
+// %operand@(1, 0): [[5, 6], [7, 8]]
+%result = "stablehlo.all_gather"(%operand) {
+  all_gather_dim = 1 : i64,
+  replica_groups = dense<[[0, 1]]> : tensor<1x2xi64>,
+  // channel_id = 0
+  channel_handle = #stablehlo.channel_handle<handle = 0, type = 0>
+  // use_global_device_ids = false
+} : (tensor<2x2xi64>) -> tensor<2x4xi64>
+// %result@(0, 0): [[1, 2, 5, 6], [3, 4, 7, 8]]
+// %result@(1, 0): [[1, 2, 5, 6], [3, 4, 7, 8]]
+```
+
+### all_reduce
+
+#### Semantics
+
+Within each process group in the StableHLO process grid, applies a reduction
+function `computation` to the values of the `operands` tensors from each process
+and produces a `results` tensors.
```

Comment:
Based on the wording from XLA semantics "When operand is a tuple of arrays, the all-reduce is performed on each element of the tuple."

```
When multiple operands are specified, the all_reduce is performed on each operand.
```

Reply:
Thanks, this will make the spec clearer.